Prompt API: Add overview for Prompt API (#43848)
mdn/content: files/en-us/web/api/languagemodel/index.md

`LanguageModel` instances cannot be constructed directly. Instead, use the static {{domxref("LanguageModel.create_static", "LanguageModel.create()")}} method.

{{InheritanceDiagram}}

## Static methods

- {{domxref("LanguageModel.availability_static", "LanguageModel.availability()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with an `Availability` value indicating whether the language model is available for the given options.
- {{domxref("LanguageModel.create_static", "LanguageModel.create()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with a new `LanguageModel` session, downloading the model if necessary.

## Instance methods

- {{domxref("LanguageModel.append()")}}
  - : Returns a {{jsxref("Promise")}} that resolves when the given input has been added to the session's context window, without generating a response.
- {{domxref("LanguageModel.clone()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with a new `LanguageModel` session that is a copy of the current session, including all context.

> - `LanguageModel.destroy()`

Reviewer: This would normally have its own page, mirroring the others. Minimally it is worth suggesting when this should be called - does it need to be called when a page is destroyed? And the implications of calling/not calling it.

Author: This appears to be a hallucination that neither of us has noticed before now. Removing.

>   - : Destroys the session and frees any associated resources. Inherited from `DestroyableModel`.

Reviewer: ```suggestion
  - : Destroys the session and frees any associated resources.
```

- {{domxref("LanguageModel.measureContextUsage()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with the number of context window tokens that the given input would consume.
- {{domxref("LanguageModel.prompt()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with the model's complete response to the given input.
- {{domxref("LanguageModel.promptStreaming()")}}
  - : Returns a {{domxref("ReadableStream")}} that streams the model's response to the given input as it is generated.

## Instance properties

- {{domxref("LanguageModel.contextUsage")}} {{ReadOnlyInline}}
  - : Returns the number of context window tokens currently consumed by this session.
- {{domxref("LanguageModel.contextWindow")}} {{ReadOnlyInline}}
  - : Returns the total context window size available for this session, in tokens.

## Events

> - {{domxref("LanguageModel.oncontextoverflow")}}

Reviewer: ```suggestion
- [`contextoverflow`](/en-US/docs/Web/API/LanguageModel/contextoverflow_event)
```

>   - : Fired when a `prompt()`, `promptStreaming()`, or `append()` call would exceed the context window size. Handled by the {{domxref("LanguageModel.oncontextoverflow")}} property.

Reviewer: ```suggestion
  - : Fired when a `prompt()`, `promptStreaming()`, or `append()` call would exceed the context window size.
```

## Examples

### Creating a session and prompting the model

> 

Reviewer: Word are needed. At this point a reader knows very little about anything so you need to provide minimal context. Pick any example on MDN as a model and follow that.

YOu could do something like

---

First we create a sessiong using the [`create()`] (LINK) method, specifying that the model should behave as a "helpful assistant" using the "system" role. 
Note that the method returns a `Promise`, so we have to `await` on it.
This may take some tiem to resolve if the model needs to be downloaded.

```js
const session = await LanguageModel.create({
  initialPrompts: [
    {
      role: "system",
      content: "You are a helpful assistant.",
    },
  ],
});
```

Once we have the model we can call `prompt()` on it to ask specific questions and log the response.

```js
const response = await session.prompt("What is the capital of France?");
console.log(response); // "The capital of France is Paris."
```

```js
const session = await LanguageModel.create({
  initialPrompts: [
    {
      role: "system",
      content: "You are a helpful assistant.",
    },
  ],
});

const response = await session.prompt("What is the capital of France?");
console.log(response); // "The capital of France is Paris."
```

### Streaming a response

> 

Reviewer: Ditto for all of these, and every example

---
title: LanguageModel
slug: Web/API/LanguageModel
page-type: web-api-interface
spec-urls: https://webmachinelearning.github.io/prompt-api/
---

{{APIRef("Prompt API")}}{{SecureContext_Header}}

The **`LanguageModel`** interface of the [Prompt API](/en-US/docs/Web/API/Prompt_API) represents a session with a browser-provided language model. It exposes static methods for creating sessions and checking availability, as well as instance methods for prompting the model, appending context, and managing the context window.

> `LanguageModel` cannot be constructed directly. Instead, use the static {{domxref("LanguageModel.create_static", "LanguageModel.create()")}} method.

Reviewer: ```suggestion
`LanguageModel` instances cannot be constructed directly. Instead, use the static {{domxref("LanguageModel.create_static", "LanguageModel.create()")}} method.
```

- {{domxref("LanguageModel.availability_static", "LanguageModel.availability()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with an `Availability` value indicating whether the language model is available for the given options.

## Instance methods

- {{domxref("LanguageModel.prompt()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with the model's complete response to the given input.
- {{domxref("LanguageModel.promptStreaming()")}}
  - : Returns a {{domxref("ReadableStream")}} that streams the model's response to the given input as it is generated.
- {{domxref("LanguageModel.append()")}}
  - : Returns a {{jsxref("Promise")}} that resolves when the given input has been added to the session's context window, without generating a response.
- {{domxref("LanguageModel.measureContextUsage()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with the number of context window tokens that the given input would consume.
- {{domxref("LanguageModel.clone()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with a new `LanguageModel` session that is a copy of the current session, including all context.
- `LanguageModel.destroy()`

>   - : Destroys the session and frees any associated resources. Inherited from `DestroyableModel`.

Reviewer: Alphabetic ordering

Author: Fixed. 

`destroy()` is a mixin. I don't remember how you handle those.

Reviewer: Generally you only document interfaces, so the information that it is in a mixin is thrown away - so just "Destroys the session and frees any associated resources." Fixed in https://github.com/mdn/content/pull/43848/changes#r3223130981

## Instance properties

- {{domxref("LanguageModel.contextUsage")}} {{ReadOnlyInline}}
  - : Returns the number of context window tokens currently consumed by this session.
- {{domxref("LanguageModel.contextWindow")}} {{ReadOnlyInline}}
  - : Returns the total context window size available for this session, in tokens.
- {{domxref("LanguageModel.oncontextoverflow")}}

>   - : An event handler for the `contextoverflow` event, which fires when a prompt would exceed the session's context window.

Reviewer: See other overviews for events. Normally are own events section.

Otherwise alphabetic for methods/properties/events etc

Reviewer: Ah, actually I see you have an events entry below. This line would be deleted, and the event would become the API entry covering this event and the event handler.

Author: Fixed.

- {{domxref("LanguageModel.destroy()")}}
  - : Releases the resources assigned to the `LanguageModel` instance it is called on and stops any further activity on it.
- {{domxref("LanguageModel.measureContextUsage()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with the number of context window tokens that the given input would consume if it were used in an operation such as `prompt()` or `append()`.
- {{domxref("LanguageModel.prompt()")}}
  - : Returns a {{jsxref("Promise")}} that resolves with the model's complete response to the given input.
- {{domxref("LanguageModel.promptStreaming()")}}
  - : Returns a {{domxref("ReadableStream")}} that streams the model's response to the given input as it is generated.

## Instance properties

- {{domxref("LanguageModel.contextUsage")}} {{ReadOnlyInline}}
  - : Returns the number of context window tokens currently consumed by this session.
- {{domxref("LanguageModel.contextWindow")}} {{ReadOnlyInline}}
  - : Returns the total context window size available for this session, in tokens.

## Events

- {{domxref("LanguageModel.contextoverflow_event", "contextoverflow")}}

>   - : Fired when a `prompt()`, `promptStreaming()`, or `append()` call would exceed the context window size.

Reviewer: ```suggestion
  - : Fired when a `prompt()`, `promptStreaming()`, or `append()` call exceeds the context window size.
```
